### ISSUE TITLE
Truncated tweets

### ISSUE THREAD

**Issue by Azenet** [oysttyer/oysttyer#83]
Since yesterday Twitter allows links and other fluff to not count towards the 140 character limit, however on oysttyer this seems to chop off some tweet ends. These tweets are displayed ending with the `…` character, and a link in the following pattern : `twitter.com/i/web/status/123456`. This seems to happen when there is an image/a quoted tweet/any link at the end of the tweet (which doesn't count towards the 140 character limit anymore). `/dump` for a truncated tweet contains a `truncated = true` property.
To be clear, this isn't about a link missing, there's text missing at the end of truncated tweets.

[Example tweet to reproduce](https://twitter.com/i/web/status/778051554549043201), or type `/th 778051554549043201`

**Comment by funnelfiasco** [comment]
Thanks for the feedback! I noticed this yesterday, too, but hadn't gotten around to looking at it. It looks like the default is to use "compatibility mode" which does the truncating. I'm working on a change that would enable the new "extended" mode.

**Comment by funnelfiasco** [comment]
Good news/bad news!

The good news is that I have a quick hack that addresses the issue.

The bad news is that somewhere along the way, it's causing a lot of error messages. It also broke the display of DMs.

So I think I'm on the right track, I just need to be a little more selective in my search and replace. Which I'll probably need to do in order to make the mode optional. I need to get back to $dayjob for now, but I'll keep working on this.

**Comment by funnelfiasco** [comment]
Okay, so this turned out to be way more frustrating than I expected. I've added a switch (`-extended`) to enable the fetching of extended tweets. The resulting JSON has the status in the `full_text` field instead of the `text` field.

I've added some logic to use `full_text` instead of `text` automatically. Both of these changes are in my WIP branch: https://github.com/oysttyer/oysttyer/tree/issue83-truncated_tweets

Here's the frustrating part: I did a find/replace, which mostly works but results in error messages as streaming events are received. So I tried a more selective approach (e.g. only updating line 6117), which results in either the initial loading of tweets to render properly and subsequent updates to only show the extra parts, or the initial loading of tweets only shows the extra parts and subsequent updates still show the truncated version.

I don't understand why the behavior changes, since AFAICT it all hits the same code path to display. I'll try working on it some more, probably tomorrow night, but I went ahead and shared my branch if anyone else wants to start poking at it some.

It also occurs to me that status length is going to be annoyingly hard to calculate now. I wonder if we're better off just getting rid of the 140-char check in oysttyer and let the API call tell us if it's too lengthy.

**Comment by Llammissar** [comment]
>  I wonder if we're better off just getting rid of the 140-char check in oysttyer and let the API call tell us if it's too lengthy.

That seems kind of okay; I was already dealing with that because of a bug in (I think) Term::Readline::TTYtter where it's off by three characters for every link.

But if we do that, is it at least possible to tell the user how many characters the API thinks it is?  Not knowing how far to trim your text is a little vexing (exacerbated because /reply and /replyall don't get expanded until you hit enter).

**Comment by atomicules** [comment]
thanks for looking into this.

I can take a look... but probably not until the weekend.

The code base is fun. Always takes a bit of warming up to follow it 
around again.

**Comment by funnelfiasco** [comment]
With help from the awesome @myshkin, I was able to come up with a version that works (https://github.com/oysttyer/oysttyer/tree/issue83-truncated_tweets), with one annoying problem:

It seems this change breaks streaming. Although oysttyer _thinks_ it is using streaming mode, the behavior is like the REST API. If you run master and this branch side-by-side, it becomes pretty clear. Something about adding the extended mode to the call seems to make cURL unhappy in ways I haven't yet been able to figure out.

One theory is that we shouldn't be passing a mode to the streaming API at all. According to https://dev.twitter.com/overview/api/upcoming-changes-to-tweets, the streaming API should already provide extended tweets. But if I take extended mode out of the arguments (~ line 4679), I get `>>> unknown server event received (non-fatal)` instead of tweets. It appears what's happening is that the tweet isn't being recognized as a tweet, so it's getting sent off to `&streamevents`, which also doesn't know what to do with it.

I've tried various edits and can't get it to handle the stream. We're frustratingly close, but not quite. I could merge my branch into master in order to get the extended tweet functionality, but I don't know if it's worth the "breaks streaming" regression.

**Comment by funnelfiasco** [comment]
(Didn't mean to close this. Obviously I'm too tired to keep working on it tonight)

**Comment by myshkin** [comment]
I think I have a fix, but it's ugly, and I haven't tested it much. Basically, we go back to the way we were before, except that when extended is on, normalizejson pushes any occurrence of full_text into text. See my latest pull request.

**Comment by funnelfiasco** [comment]
I tested @myshkin's fix and it works, so it's merged into my branch, which I'm about to merge to master. I've reconsidered removing the limit calculation since it could still be useful in most cases. We could add the logic to count correctly, but that's....not pretty. Setting `linelength` to a large number is a suitable way to disable it for now.

**Comment by myshkin** [comment]
Argh, we're not quite done. In streaming mode, but not via /th, (some?) quoted tweets are showing up as blank with HEAD of master. The original tweet looks fine, but the quoted tweet just has a key and handle, with no text displayed.

**Comment by funnelfiasco** [comment]
@myshkin I haven't noticed it yet, but I've been AFK much of the weekend. Do you have any examples you can share? It sounds like it's not consistent, so we might have to comb through the JSON to find the common settings.

**Comment by myshkin** [comment]
Unfortunately, no, and I've seen a number of fine quoted tweets in the interim. But just now, I got one that looked fine in streaming mode, but with "/a mikeduncan", the tweet with ID 782711946365722624 quotes @BBCBreaking, and the quoted tweet shows up as blank. A dump indicates that full_text has the proper value, but text is empty. I guess quote tweets need to get normalized as well.

I do also see some instances in which t.co links don't get destroyed, e.g. ID 782708874407190529. There must be a better way to rewrite the flow of normalizejson().

**Comment by funnelfiasco** [comment]
Hmm, yep. Looks like `/again`-ing doesn't get the tweet printed (though I see it in the stream). I'll take a look.

**Comment by funnelfiasco** [comment]
I have a fix in place that I'm testing. It seems to work, but I'll hold off on pushing it for a little bit just to be sure. Spoiler alert: it will make you cry.

**Comment by funnelfiasco** [comment]
Fixed the t.co issue, I think, in the commit I just pushed. As it turns out, we were only replacing t.co links in `text`, so I added replacements for `full_text`, too.

Discovered that my "fix" for getting statuses to render caused them to be truncated (which I suppose makes sense given what I did). Still working on it when I have spare moments.

**Comment by myshkin** [comment]
It's not clear to me why `destroy_all_tco()` should need to modify `full_text`; the point of my last commit was that nearly all of the code should only have to deal with `text`. Maybe `destroy_all_tco()` is too early in the rendering process?

My tears are ready to go.

**Comment by myshkin** [comment]
I got the same tweet (783035893766623232) twice, once from /a and once from streaming. Running /th on the different menu codes for each, the /a one appears fine, and the streaming one is blank. The retweeted tweet looks fine in both cases. There's something inconsistent about how `retweeted_status` gets populated.

**Comment by funnelfiasco** [comment]
Yeah, I think `destroy_all_tco()` ends up upstream of most stuff, since it's also used in profiles, etc., etc.
I'm tempted to say "some cases will just stay truncated for now and we'll fix that later".

**Comment by myshkin** [comment]
Throwing a bunch of `grabjson()` calls into `normalizejson()` seems to do the trick, as now the RT and QT structures are fully populated. Is there a more efficient way to do that?

**Comment by funnelfiasco** [comment]
Oh hell, my change to destroy_all_tco seems to make life worse (RTs don't get rendered). Not sure how I missed that. I disappeared my changes from today because it turns out they're awful. Can you open a pull request against master as it currently stands?

**Comment by myshkin** [comment]
Done. (Took me a while because I had already merged your changes into my local copy of master.)

**Comment by funnelfiasco** [comment]
Okay, it looks like we have a working solution thanks to @myshkin.
